posydon/binary_evol/pulsar_modeling.py

```diff
@@ -0,0 +1,256 @@
+        self.moment_inertia = self.calc_moment_of_inertia()   ## moment of inertia of the NS
+        ## Note: Because the NS radius is constant, moment of inertia should also be constant throughout NS evolution
+
+        self.spin = self.draw_NS_spin()          ## NS spin angular frequency [1/s]
+        self.Bfield = self.draw_NS_Bfield()      ## NS magnetic field [G]
+
+        self.alive_state = self.is_alive()
+ 
+        
+    def draw_NS_spin(self):
+        '''
+        Draw the initial NS spin angular frequency Omega [1/s] from a uniform random distribution.
+        Range is 0.1-2.0 sec for pulse period P.
+        Omega = 2*pi/P
+        '''
+        return np.random.uniform(2*np.pi/.1, 2*np.pi/2) 
+
+    def draw_NS_Bfield(self):
+        '''
+        Draw the initial NS B-field [G] from a lognormal distribution.
+        Range is 10^11.5 - 10^13.8 Gauss.
+        '''
+        return np.random.lognormal(3.16e11, 6.31e13)
+    
+    def calc_moment_of_inertia(self):
+        '''
+        Calculate the moment of intertia of the NS in g*cm^2
+        Eq. from Kiel et al. 2008 is in solar units
+        '''
+        M = self.mass/const.Msun       ## mass of the NS [Msun] 
+        R = self.radius/const.Rsun     ## radius of the NS [Rsun]
+
+        return 2/7 * (1 - 2.42e-6*M/R - 2.9e-12*M**2/R**2)**-1 * M*R**2 * (const.Msun*const.Rsun**2)
+    
+    def calc_NS_edd_lim(self, surface_h1):
+        """
+        Get Eddington accretion rate for a NS.
+        Adapted from eddington_limit() in utils.common_functions
+
+        Parameters
+        ----------
+        NS_mass: mass of the NS [g]
+        surface_h1: surface Hydrogen fraction of the NS??? ask if this makes sense
+        acc_radius: radius of the NS [cm]
+
+        """
+        #surface_h1 = 0
+        eta = const.standard_cgrav * self.mass/(self.radius * const.clight**2)
+        Mdot_edd = (4 * np.pi * const.standard_cgrav * self.mass * (0.2 * (1 + surface_h1) * eta * const.clight))   
+        return Mdot_edd
+    
+    def detached_evolve(self, delta_t):
+        '''
+        Evolve a pulsar during detached evolution.
+
+        Parameters
+        ----------
+        delta_t: the duration of the detached evolution phase [s]
+        '''
+        alpha = 45*const.a2rad          ## angle between the axis of rotation and magnetic axis [rad]
+        mu_0 = 1                        ## permeability of free space [has value unity in cgs]
+        B_min = 1e8                     ## minimum Bfield strength at which Bfield decay ceases [G]
+        tau_d = 3*1e9*const.secyer      ## B-field decay timescale = 3 Gyr [s]
+
+        delta_t = delta_t*const.secyer
+
+        I = self.moment_inertia
+        R = self.radius
+        B_i = self.Bfield
+
+        ## evolve the NS B-field
+        B_f = (B_i - B_min) * np.exp(-delta_t/tau_d) + B_min
+        self.Bfield = B_f
+
+        ## evolve the NS spin
+        A = 8*np.pi*R**6*np.sin(alpha)**2/(3*mu_0*const.clight**3*I)
```

Comment:
Shouldn't there be a negative sign here for A in line 126? Unless you took care of it in the equation for final spin.